### PR TITLE
Add `Locale` conversion format configuration property

### LINKED ISSUE #3141: `Locale` argument conversion not setting `language` and `country` properly
I am trying to write a parameterized test for a service that uses a resource bundle under the hood, loading the resource bundle with [`ResourceBundle.getBundle(String, Locale)`](https://docs.oracle.com/javase/8/docs/api/java/util/ResourceBundle.html#getBundle-java.lang.String-java.util.Locale-).

The test gets several `Locale` values in a `@CsvSource` annotation: 

The [documentation](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests-argument-conversion-implicit) does not have an example for a country-based locale so I initially tried with:

```java
@CsvSource({
  "en_FR, value1",
  "en_GB, value2",
  "fr_FR, value3",
  "fr_MC, value4",
  "it_IT, value5",
  ...
})
void test(Locale locale, String expected) {
  // call service using the resource bundle

  // compare result with `expected`
}
```

All worked fine on Windows (local environment) but failed on Linux (CI).

After some digging, on Windows everything works because it's a case-insensitive file system, so for example the `en_FR` value is converted to a `en_fr` language-only locale (i.e., without country) and it successfully matches the file ending with `en_FR` because of case insensitivity.

Looking at the code, I think the problem is here:

https://github.com/junit-team/junit5/blob/4288cf1233b4af6002e922cb9ddd1b159fe2c9b3/junit-jupiter-params/src/main/java/org/junit/jupiter/params/converter/DefaultArgumentConverter.java#L265

which uses the [`Locale(String)`](https://docs.oracle.com/javase/8/docs/api/java/util/Locale.html#Locale-java.lang.String-) constructor. The constructor Javadoc also explains the different behavior based on case sensitivity:

> This constructor normalizes the language value to lowercase. 

Looking at the test cases:

https://github.com/junit-team/junit5/blob/4288cf1233b4af6002e922cb9ddd1b159fe2c9b3/junit-jupiter-params/src/test/java/org/junit/jupiter/params/converter/DefaultArgumentConverterTests.java#L222-L223

Probably the second one is invalid. I would have expected that to be:

```java
assertConverts("en_us", Locale.class, Locale.US);
```

which does not pass with the current implementation.

## Steps to reproduce

As the documentation does not cover conversion of locales with country, I am not sure what would be the "right" test case to demonstrate the issue.

Assuming the [IETF BCP 47 language tag format](https://docs.oracle.com/javase/8/docs/api/java/util/Locale.html#forLanguageTag-java.lang.String-) would be the right format for input values, this test fails already at the first assertion:

```java
  @ParameterizedTest
  @ValueSource(strings = { "en-US" })
  void test(Locale locale) {
    assertEquals("en", locale.getLanguage());
    assertEquals("US", locale.getCountry());
    assertEquals(Locale.US, locale);
  }
```

## Workaround

Currently, I'm using a custom `ArgumentConverter` which delegates the conversion to [`Locale::forLanguageTag`](https://docs.oracle.com/javase/8/docs/api/java/util/Locale.html#forLanguageTag-java.lang.String-).

## Context

 - Used versions (Jupiter/Vintage/Platform): 5.8.2
 - Build Tool/IDE: Intellij IDEA / Maven

## Deliverables

- [ ] `DefaultArgumentConverter` properly converts IETF BCP 47 strings
- [ ] The [documentation](https://junit.org/junit5/docs/current/user-guide/#writing-tests-parameterized-tests-argument-conversion-implicit) has an example using the IETF BCP 47 language tag format

### PR DESCRIPTION
**Pull request by scordio**
## Overview

Fixes #3141.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [ ] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [ ] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [ ] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [ ] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)

### CODE REVIEW

File: junit-jupiter-engine/junit-jupiter-engine.gradle.kts [hunk 1]
```diff
@@ -13,6 +13,7 @@ dependencies {
 	api(platform(projects.junitBom))
 	api(projects.junitPlatformEngine)
 	api(projects.junitJupiterApi)
+	api(projects.junitJupiterParams)
```
Comment (scordio) [inline junit-jupiter-engine/junit-jupiter-engine.gradle.kts:16]: Looking for Gradle/Module System help 🙏

What else am I missing to solve the following? Triggered by `junit-jupiter-engine:compileModule`:

```
> Task :junit-jupiter-engine:compileModule FAILED
/Users/stefano/Projects/junit5/junit-jupiter-engine/build/module/org.junit.jupiter.engine/module-info.java:23: error: module not found: org.junit.jupiter.params.converter
	requires org.junit.jupiter.params.converter;
	                                 ^
1 error
```

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/converter/LocaleConversionFormat.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import static org.apiguardian.api.API.Status.INTERNAL;
+
+import org.apiguardian.api.API;
+
+/**
+ * Enumeration of {@link java.util.Locale} conversion formats.
+ *
+ * @since 5.11
+ */
+@API(status = INTERNAL, since = "5.11")
```
Comment (scordio) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/converter/LocaleConversionFormat.java:22]: Is it correct to mark this as an internal API as the users are supposed to set these values in the config property only and not rely on the enum?

File: junit-jupiter-engine/junit-jupiter-engine.gradle.kts (hunk 1 above)
Comment (scordio) [reply junit-jupiter-engine/junit-jupiter-engine.gradle.kts:16]: Maybe I could summon @sormuras for a hint? 😇

**Comment by scordio** [comment]
Waiting for #3507 to be finalized before moving forward.

**Comment by marcphilipp** [comment]
@scordio Are you still interested in continuing with this PR?

**Comment by scordio** [comment]
Yes! This has fallen under my radar... thanks for the reminder 🙂

I'll rebase it and catch up with the missing parts in the upcoming weeks.

**Comment by scordio** [comment]
Superseded by #4219.

**Comment by marcphilipp** [comment]
Yes, I think we should! :+1:

I couldn't reopen it, though, since you have already deleted the underlying branch. Would you like to restore it or rather start a new PR?

**Comment by scordio** [comment]
Although I still had the branch locally, I couldn't trick GitHub to reopen this PR 😅

I'll start fresh then!

**Comment by scordio** [comment]
Superseded by #4492.

**Pull request closed without merging**